README.md
# tft-suite

## Description
tft-suite is an active development library for creating and switching between different screens
on an Adafruit Mini PiTFT. Adafruit offers these displays in two sizes [(135x240)](https://www.adafruit.com/product/4393) and [(240x240)](https://www.adafruit.com/product/4484).
Both of these displays come with two small tactile buttons allowing users to interact with a raspberry pi or the display.

## Table of Contents
1. [Installation](#installation)
2. [Usage](#usage)
3. [License](#license)

## Installation

```shell
sudo apt-get install ttf-dejavu
python -m pip install git+https://github.com/drewtchrist/tft-suite
```

## Usage
tft-suite offers a small handful of screens that should work right out of the box. Some of the
screens have code that has been adapted from AdaFruit's [guide here](https://learn.adafruit.com/adafruit-mini-pitft-135x240-color-tft-add-on-for-raspberry-pi),
such as the one that shows stats about the raspberry pi. To use one of the screens that comes 
with tft-suite, import the Display class and any screens you would like to use like below.

```python
from tft_suite.display import Display
from tft_suite.screens.stats_screen import StatsScreen

if __name__ == '__main__':
    Display((135, 240), [
        (StatsScreen, ())
    ]).start()
```

Alternatively, anyone can create their own custom Screen class if they'd like to 
display something that tft-suite doesn't offer! This process should be simplified
later on, but the example below should be enough to get someone started. 

```python
from tft_suite.screens.screen import Screen

class HelloWorldScreen(Screen):

    def __init__(self, **kwargs):
        super(HelloWorldScreen, self).__init__(**kwargs)

    def draw_screen(self):
        self.draw.rectangle((0, 0, self.width, self.height), outline=0, fill=0)
        self.draw.text((0, 0), "Hello, World!", font=self.font, fill="#FF00FF")
        self.display.image(self.image)
```

After your screen is created, import it and add it to the list of Screens in your
Display class like the first example.

![tft-suite gif](https://drive.google.com/uc?export=view&id=1BJnZUrAmg03JISMqwGYimUwBA3INzrGA)

When the code is running, screens can be cycled through with either of the two buttons.
The backlight can be turned off by quickly pressing both buttons at the same time.

## License



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added

### Changed

### Removed

## [0.3.2](https://github.com/drewtchrist/tft-suite/releases/tag/v0.3.2) - 2021-10-31
### Added

### Changed
* Updated readme example... yes this needed released

### Removed

## [0.3.0](https://github.com/drewtchrist/tft-suite/releases/tag/v0.3.0) - 2021-10-31

## [0.3.1](https://github.com/drewtchrist/tft-suite/releases/tag/v0.3.1) - 2021-10-31
### Added

### Changed
* Updated readme example... yes this needed released

### Removed

## [0.3.0](https://github.com/drewtchrist/tft-suite/releases/tag/v0.3.0) - 2021-10-31
### Added
* 240x240 support
* Typed the project

### Changed
* Disabling the backlight now stops the screenthread to save those precious cycles

### Removed

## [0.2.0](https://github.com/drewtchrist/tft-suite/releases/tag/v0.2.0) - 2021-10-16
### Added
* Added an abstract method `draw_screen` to the `Screen` class

### Changed
* The `run` method of the `Screen` class is no longer abstract and calls the `draw_screen` method

### Removed



tft_suite/display.py
import time

import digitalio
import board
import adafruit_rgb_display.st7789 as st7789
from PIL import Image, ImageDraw, ImageFont
from tft_suite.switcher import Switcher

class Display:
    """Display class representing an Adafruit Mini PiTFT"""

    def __init__(self, size, screens=None):

        # hardware
        self.display = None
        self.backlight = None
        self.buttonA = None
        self.buttonB = None

        # size
        self.width = None
        self.height = None

        # drawables
        self.image = None
        self.draw = None

        # screens
        self.screens = [] if not screens else screens

        self.initialize_hardware(size)
        self.initialize_drawables()

        self.switcher = Switcher(
            self.screens,
            **{
                'height': self.height,
                'width': self.width,
                'draw': self.draw,
                'image': self.image,
                'display': self.display
            }
        )

    def create_display(self, size):
        if size[0] == 135:
            return st7789.ST7789(
                board.SPI(),
                cs=digitalio.DigitalInOut(board.CE0),
                dc=digitalio.DigitalInOut(board.D25),
                rst=None,
                baudrate=64000000,  # The pi can be very fast!
                width=size[0],
                height=size[1],
                x_offset=53,
                y_offset=40,
                rotation=90
            )
        elif size[0] == 240:
            return st7789.ST7789(
                board.SPI(),
                cs=digitalio.DigitalInOut(board.CE0),
                dc=digitalio.DigitalInOut(board.D25),
                rst=None,
                baudrate=64000000,  # The pi can be very fast!
                width=size[0],
                height=size[1],
                y_offset=80,
                rotation=180
            )


    def initialize_hardware(self, size):
        # Create the ST7789 display:
        self.display = self.create_display(size)
        self.backlight = digitalio.DigitalInOut(board.D22)
        self.backlight.switch_to_output()
        self.backlight.value = True

        self.buttonA = digitalio.DigitalInOut(board.D24)
        self.buttonB = digitalio.DigitalInOut(board.D23)

        self.buttonA.switch_to_input()
        self.buttonB.switch_to_input()


    def initialize_drawables(self):
        # flip these because the display is horizontal
        self.height = self.display.width
        self.width = self.display.height

        self.image = Image.new("RGB", (self.width, self.height))

        self.draw = ImageDraw.Draw(self.image)
        self.display.image(self.image)


    def enable_backlight(self):
        self.backlight.value = True
        

    def toggle_backlight(self):
        self.backlight.value = not self.backlight.value


    def buttonA_pressed(self):
        return self.buttonA.value and not self.buttonB.value


    def buttonB_pressed(self):
        return self.buttonB.value and not self.buttonA.value


    def buttons_pressed(self):
        return not self.buttonA.value and not self.buttonB.value


    def start(self):
        self.switcher.start()
        while True:
            try:
                # A button is pushed
                if self.buttonA_pressed():
                    if self.switcher.running:
                        self.switcher.switch_up()
                        time.sleep(0.5)
                    #else:
                    #    print('buttonA switcher is not running')

                # B button is pushed
                if self.buttonB_pressed():
                    if self.switcher.running:
                        self.switcher.switch_down()
                        time.sleep(0.5)
                    #else:
                    #    print('buttonB switcher is not running')

                # Both buttons are pushed
                if self.buttons_pressed():
                    self.toggle_backlight()
                    time.sleep(0.5)
                    #print(self.switcher.screenthread.is_alive())
                    #print(f'switcher running: {self.switcher.running}')
                    if self.switcher.running:
                        self.switcher.stop()
                        #print('switcher stopped')
                    else:
                        self.switcher.start()
                        #print('switcher started')

                    time.sleep(0.5)
                
            except KeyboardInterrupt:
                self.switcher.stop()
                exit()


